### PR TITLE
Add xxhash64 support

### PR DESCRIPTION
**Pull request by nvdbaranec**
Adds support for xxhash64.  Includes JNI bindings, cpp tests and java tests.  

Does not currently support structs or lists, but those should be relatively easy to add if desired.

<s>Dependent on https://github.com/NVIDIA/spark-rapids-jni/pull/1246</s>

### CODE REVIEW

File: src/test/java/com/nvidia/spark/rapids/jni/HashTest.java [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+    // The hash values were derived from Apache Spark in a manner similar to the one documented at
+    // https://github.com/rapidsai/cudf/blob/aa7ca46dcd9e/cpp/tests/hashing/hash_test.cpp#L281-L307
+    try (ColumnVector v = ColumnVector.decimalFromInts(-3,
+        0, 100, -100, 0x12345678, -0x12345678);
+         ColumnVector result = Hash.xxhash64(new ColumnVector[]{v});
+         ColumnVector expected = ColumnVector.fromBoxedLongs(-5252525462095825812L, 8713583529807266080L, 5675770457807661948L, -958054811, -1447702630)) {
+      assertColumnsAreEqual(expected, result);
+    }
+  }
+  */
```
Comment (jlowe) [inline src/test/java/com/nvidia/spark/rapids/jni/HashTest.java:339]:
```suggestion

  @Test
  void testXXHash64Decimal32() {
    // The hash values were derived from Apache Spark in a manner similar to the one documented at
    // https://github.com/rapidsai/cudf/blob/aa7ca46dcd9e/cpp/tests/hashing/hash_test.cpp#L281-L307
    try (ColumnVector v = ColumnVector.decimalFromInts(-3,
        0, 100, -100, 0x12345678, -0x12345678);
         ColumnVector result = Hash.xxhash64(new ColumnVector[]{v});
         ColumnVector expected = ColumnVector.fromBoxedLongs(-5252525462095825812, 8713583529807266080, 5675770457807661948, -7728554078125612835, 3142315292375031143)) {
      assertColumnsAreEqual(expected, result);
    }
  }
```

**Review by jlowe — COMMENTED** [review]
Marking as draft since this depends on #1244 and #1246.  This also needs a signoff commit per the contribution guidelines.

File: src/main/java/com/nvidia/spark/rapids/jni/Hash.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.jni;
+
+import ai.rapids.cudf.*;
```
Comment (jlowe) [inline src/main/java/com/nvidia/spark/rapids/jni/Hash.java:19]: Nit: This shouldn't include _everything_ from cudf.  There's very little that needs to be pulled into cudf here.

File: src/main/cpp/src/HashJni.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+extern "C" {
+
+JNIEXPORT jlong JNICALL Java_com_nvidia_spark_rapids_jni_Hash_murmurHash32(
+  JNIEnv* env, jobject j_object, jint seed, jlongArray column_handles)
+{
+  JNI_NULL_CHECK(env, column_handles, "array of column handles is null", 0);
+
+  try {
+    auto column_views =
```
Comment (jlowe) [inline src/main/cpp/src/HashJni.cpp:31]:
```suggestion
    cudf::jni::auto_set_dwevice(env);
    auto column_views =
```

File: src/main/cpp/src/HashJni.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+JNIEXPORT jlong JNICALL Java_com_nvidia_spark_rapids_jni_Hash_xxhash64(JNIEnv* env,
+                                                                       jobject j_object,
+                                                                       jlong seed,
+                                                                       jlongArray column_handles)
+{
+  JNI_NULL_CHECK(env, column_handles, "array of column handles is null", 0);
+
+  try {
+    auto column_views =
```
Comment (jlowe) [inline src/main/cpp/src/HashJni.cpp:47]:
```suggestion
    cudf::jni::auto_set_device(env);
    auto column_views =
```

File: src/main/cpp/src/HashJni.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+#include "cudf_jni_apis.hpp"
+#include "dtype_utils.hpp"
+#include "jni_utils.hpp"
+
+#include "hash.cuh"
+
+extern "C" {
+
+JNIEXPORT jlong JNICALL Java_com_nvidia_spark_rapids_jni_Hash_murmurHash32(
+  JNIEnv* env, jobject j_object, jint seed, jlongArray column_handles)
```
Comment (jlowe) [inline src/main/cpp/src/HashJni.cpp:26]:
```suggestion
  JNIEnv* env, jclass, jint seed, jlongArray column_handles)
```

File: src/main/cpp/src/HashJni.cpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    auto column_views =
+      cudf::jni::native_jpointerArray<cudf::column_view>{env, column_handles}.get_dereferenced();
+    return cudf::jni::release_as_jlong(
+      spark_rapids_jni::murmur_hash3_32(cudf::table_view{column_views}, seed));
+  }
+  CATCH_STD(env, 0);
+}
+
+JNIEXPORT jlong JNICALL Java_com_nvidia_spark_rapids_jni_Hash_xxhash64(JNIEnv* env,
+                                                                       jobject j_object,
```
Comment (jlowe) [inline src/main/cpp/src/HashJni.cpp:40]:
```suggestion
                                                                       jclass,
```

**Review by jlowe — COMMENTED** [review]
lgtm, just need to get the dependencies merged first.

File: pom.xml [hunk 7]
```diff
@@ -195,20 +195,9 @@
                   <excludes>
                     <exclude>**/CuFileTest.java</exclude>
                     <exclude>**/CudaFatalTest.java</exclude>
-                    <exclude>**/ColumnViewNonEmptyNullsTest.java</exclude>
```
Comment (jlowe) [inline pom.xml:198]: Curious why pom.xml is being changed here?  This is unrelated to xxhash64 changes and effectively reverts #1252.

**Review by jlowe — CHANGES_REQUESTED** [review]

File: thirdparty/cudf
[the remark this answers is not in the document]
Comment (nvdbaranec) [reply thirdparty/cudf:1]: I'm not sure what happened here.  Probably an issue stemming from merging in 23.08 after the murmur hash move PR got merged.  I'll fix.

**Review by jlowe — APPROVED** [review]

File: src/main/cpp/src/xxhash64.cu [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+  result_type __device__ inline operator()(Key const& key) const { return compute(key); }
+
+  template <typename T>
+  result_type __device__ inline compute(T const& key) const
+  {
+    return compute_bytes(reinterpret_cast<std::byte const*>(&key), sizeof(T));
+  }
+
+  result_type __device__ inline compute_remaining_bytes(std::byte const* data,
+                                                        cudf::size_type nbytes,
```
Comment (hyperbolic2346) [inline src/main/cpp/src/xxhash64.cu:76]:
```suggestion
                                                        cudf::size_type const nbytes,
```

File: src/test/java/com/nvidia/spark/rapids/jni/HashTest.java [hunk 9]
```diff
@@ -261,4 +261,131 @@ void testSpark32BitMurmur3HashListsAndNestedLists() {
...
+  @Test
+  void testXXHash64Strings() {
+    try (ColumnVector v0 = ColumnVector.fromStrings(
+           "a", "B\nc",  "dE\"\u0100\t\u0101 \ud720\ud721\\Fg2\'",
+           "A very long (greater than 128 bytes/char string) to test a multi hash-step data point " +
+           "in the MD5 hash function. This string needed to be longer.A 60 character string to " +
+           "test MD5's message padding algorithm",
+           "hiJ\ud720\ud721\ud720\ud721", null);
+         ColumnVector result = Hash.xxhash64(new ColumnVector[]{v0});
+         ColumnVector expected = ColumnVector.fromBoxedLongs(-8582455328737087284L, 2221214721321197934L, 5798966295358745941L, -4834097201550955483L, -3782648123388245694L, Hash.DEFAULT_XXHASH64_SEED)) {
```
Comment (hyperbolic2346) [inline src/test/java/com/nvidia/spark/rapids/jni/HashTest.java:274]: Is there a reason we don't use the CPU to hash the same ColumnVector and compare?

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]
Good work on this. I appreciate the tests added for murmur as well. A few nits and questions.

File: src/test/java/com/nvidia/spark/rapids/jni/HashTest.java (hunk 9 above)
Comment (nvdbaranec) [reply src/test/java/com/nvidia/spark/rapids/jni/HashTest.java:274]: Not particularly.  Just following what the murmur hash tests do.

**Review by hyperbolic2346 — APPROVED** [review]

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

File: src/main/cpp/src/xxhash64.cu [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+  constexpr XXHash_64() = delete;
+  constexpr XXHash_64(hash_value_type seed) : m_seed(seed) {}
+
+  template <typename T>
+  __device__ inline T getblock32(std::byte const* data, cudf::size_type offset) const
+  {
+    // Read a 4-byte value from the data pointer as individual bytes for safe
+    // unaligned access (very likely for string types).
+    auto block      = reinterpret_cast<std::uint8_t const*>(data + offset);
+    uint32_t result = static_cast<uint32_t>(block[0]) | (static_cast<uint32_t>(block[1]) << 8) |
```
Comment (rwlee) [inline src/main/cpp/src/xxhash64.cu:55]: Is there a functional difference that I'm missing between this and https://github.com/NVIDIA/spark-rapids-jni/pull/1246/files#diff-3b5fbfa28ce0798e562307d27b7636b7fd46e6d90d92e93c61a0da5411aec0dbR55. The addition of the extra static I believe is unnecessary.

**Pull request merged**